Refactor array types encoding

dario23 opened the pull request:
- Moved to `Encoder` so it can be used from pure function encoder as well.
 - Moved encoding of `lookup_pure` calls onto the `EncodedArrayTypes` and `EncodedSliceTypes` structs to avoid repeating it everywhere.

dario23 committed:
Move encode_array_types, encode_slice_types to separate array_encoder

This way we avoid making the encoder.rs file even longer

dario23 committed:
Move encoding of lookup_pure to Encoded{Array,Slice}Types struct

dario23 committed:
Allow pre-encoding sequence element snapshot type

This is useful when calling `encode_array_types` from e.g. inside the
`SnapshotEncoder` where `Encoder`'s `snapshot_encoder` field is already
borrowed, so `encoder.encode_snapshot_type` would result in a double
borrow (i.e. panic).

dario23 committed:
Add caching to array_types_encoder

Aurel300 commented on prusti-viper/src/encoder/array_encoder.rs:
```diff
@@ -0,0 +1,237 @@
+    pub elem_value_ty: vir::Type,
+    pub elem_ty_rs: ty::Ty<'tcx>,
+    pub array_len: usize,
+    pub lookup_pure_name: String,
```
I wonder: can `lookup_pure` (and others) always just have a constant name? E.g. for snapshots the snapshot function is called `snap$` for any (non-primitive etc) type. Prusti later does mangling based on the types found in the signature so the function name becomes unique again.

dario23 replied:
i had the same thought some time ago, but i think it's at least not as simple as just dropping them from the name. currently we have
```
function Array$5$i32$lookup_pure__$TY$__Array$5$i32$$int$$$int$(self: Ref, idx: Int): Int
  requires acc(Array$5$i32(self), read$())
  requires idx < 5
```
so the different `lookup_pure` instances are actually different because they mention their respective type predicate and length in their preconditions. we might be able to get rid of the preconditions and sort-of inline them as assertions whenever we use them, but at that point i'm not sure it's worth it, so i dropped the idea at the time.

Aurel300 replied:
The functions would still be different, they would all just share the same name. In the example you have, everything after `__$TY$__` is not really part of the name, it is added by name mangling later.

dario23 replied:
ah, that sounds like a plan, yes. i'll change that.

Aurel300 commented on prusti-viper/src/encoder/array_encoder.rs:
```diff
@@ -0,0 +1,237 @@
+    pub elem_value_ty: vir::Type,
+    pub elem_ty_rs: ty::Ty<'tcx>,
+    pub lookup_pure_name: String,
+    pub slice_len_name: String,
```
(same comment, can these always be named the same)

Aurel300 commented on prusti-viper/src/encoder/array_encoder.rs:
```diff
@@ -0,0 +1,237 @@
+};
+
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct EncodedArrayTypes<'tcx> {
```
At least a short comment on what these represent would be useful. (same for the slice type)

Aurel300 commented on prusti-viper/src/encoder/array_encoder.rs:
```diff
@@ -0,0 +1,237 @@
+    /// Optionally pass a pre-encoded `vir::Type` for the element snapshot type. This is useful
+    /// when calling `encode_slice_types` from e.g. inside the `SnapshotEncoder` where `Encoder`'s
+    /// `snapshot_encoder` field is already borrowed, so `encoder.encode_snapshot_type` would
+    /// result in a double borrow (i.e. panic).
```
I find this `pre_encoded_elem_snap_ty` API a bit messy. How does `SnapshotEncoder` call this?

dario23 replied:
it is messy indeed. one alternative i thought about is to drop that field from the struct, but we need it in `encode_lookup_pure_call`.

SnapshotEncoder is (sometimes/mostly) called through [`Encoder::encode_snapshot_type`](https://github.com/viperproject/prusti-dev/blob/master/prusti-viper/src/encoder/encoder.rs#L881), which does `        self.snapshot_encoder.borrow_mut().encode_type(self, ty)`. The `SnapshotEncoder::encode_type` call goes through pretty much the whole snapshot machinery for the type, building the required functions and axioms and whatnot, and in the end only returning the type of the snapshot. If anything inside that machinery calls `encode_array_types`, then that would arrive at the same `self.snapshot_encoder.borrow_mut()` again, making prusti panic. So instead, `SnapshotEncoder` would just call `self.encode_type` on the array element type first, and use the result of that, without going through `Encoder::encode_snapshot_type` thus avoiding the borrowing problem.

I'm not sure if we can make that `borrow_mut` more fine-grained by moving the `Rc` into `SnapshotEncoder` (so we'd only need a shared reference for the calls from `Encoder`, or if that would just move the problem elsewhere. i hope it would drastically shorten the duration (and split it up by its fields) of the `SnapshotEncoder` are borrowed).

dario23 replied:
hm, now that i think about it.. we could make it a little less messy by dropping the element snapshot type from the `EncodedArrayTypes` struct and requiring the user to pass it to `EncodedArrayTypes::encode_lookup_pure_call`. That way, we can still pass whatever is appropriate/possible in the current situation, but we don't need the admittedly ugly API to thread this additional (not really intuitive) parameter through.

dario23 replied:
we still need it to call `encode_builtin_function_use`, which i think should be called in `encode_array_types`, as we don't have access to `encoder` anymore in `encode_lookup_pure_call` (which only has the `EncodedArrayTypes` as `self`).

i have removed all the weird "passing in `elem_value_ty`" stuff for now, and just call `encoder.encode_snapshot_type`. for how we'll do that from within `SnapshotEncoder` i'll worry later. one step at a time i guess. let's get this PR and #489 merged and take it from there.

Aurel300 commented on prusti-viper/src/encoder/array_encoder.rs:
```diff
@@ -0,0 +1,237 @@
+            }
+        );
+
+        let encoded = EncodedSliceTypes{
```
```suggestion
        let encoded = EncodedSliceTypes {
```

Aurel300 commented on prusti-viper/src/encoder/array_encoder.rs:
```diff
@@ -0,0 +1,237 @@
+            }
+        );
+
+        let encoded = EncodedArrayTypes{
```
```suggestion
        let encoded = EncodedArrayTypes {
```

Aurel300 commented on prusti-viper/src/encoder/procedure_encoder.rs:
```diff
@@ -2224,25 +2205,9 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
-                elem_ty_pred: st.elem_pred,
-            }
-        );
+        let st = self.encoder.encode_slice_types(slice_ty, None).with_span(span)?;
```
Please replace `st` and `at` with a more descriptive name. We know what they are now but … :)

dario23 committed:
Address reviews

dario23 committed:
Add docstrings for EncodedArrayTypes, EncodedSliceTypes

dario23 committed:
Don't encode type in lookup_pure name

it'll be added later anyway as part of the argument types, so we'll just
go with "lookup_pure" before mangling
